mono/metadata/pal-icalls.c

```diff
@@ -0,0 +1,102 @@
+/**
+ * \file
+ * System.Native PAL internal calls
+ *
+ * Copyright 2018 Microsoft
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+#include <config.h>
+#include <glib.h>
+#include <mono/metadata/handle.h>
+#include <mono/utils/hazard-pointer.h>
+#include <mono/utils/mono-threads.h>
+
+#include "pal-icalls.h"
+
+#if !defined(HOST_WIN32) && !defined(HOST_WASM) /* Unix System.Native PAL */
+
+extern int32_t SystemNative_Read(intptr_t fd, void* buffer, int32_t bufferSize);
+
+static void
+interrupt_read (gpointer data)
+{
+	MonoNativeThreadId tid = (MonoNativeThreadId) data;
+	MonoThreadHazardPointers *hp;
+	MonoThreadInfo *info;
+
+	if (tid == mono_native_thread_id_get ())
+		return;
+
+	mono_thread_info_suspend_lock ();
+	hp = mono_hazard_pointer_get ();
+	info = mono_thread_info_lookup (tid);
+	if (!info) {
+		mono_thread_info_suspend_unlock ();
+		return;
+	}
+	mono_threads_begin_global_suspend ();
+	mono_threads_suspend_abort_syscall (info);
+	mono_threads_wait_pending_operations ();
+
+	mono_hazard_pointer_clear (hp, 1);
+	mono_threads_end_global_suspend ();
+	mono_thread_info_suspend_unlock ();
+}
+
+gint32
+ves_icall_CoreFX_Interop_Sys_Read (MonoSafeHandleHandle fd_handle, gchar* buffer, gint32 count)
+{
+	gint32 result;
+	gpointer fd = MONO_HANDLE_GETVAL(fd_handle, handle);
+	gboolean interrupted;
+
+	do {
+		mono_thread_info_install_interrupt (interrupt_read, (gpointer) (gsize) mono_native_thread_id_get (), &interrupted);
```

Comment:
This `install`+`uninstall` is not needed since a signal will interrupt the syscall and that's what you are trying to do in `interrupt_read`. Why not doing the `while` loop in managed instead?

Reply:
@luhenry I don't understand this point. We want the runtime itself to be able to interrupt the read, so that it can shut down a thread blocked in `read`. If the runtime doesn't know to do this, it will not be interrupted and the process will hang on exit. The purpose of using an icall for this was to allow this to happen.

Reply:
The runtime already tries to abort the thread, even if there is no `install`+`uninstall` around the blocking syscall, so what you are doing in `interrupt_read` is redundant with what the runtime does regardless of the presence of an interrupt token or not. Or in other words: it's not because you don't install an interrupt callback that the runtime won't send an abort signal (that's what `mono_threads_suspend_abort_syscall` does).

Reply:
> The runtime already tries to abort the thread, even if there is no install+uninstall around the bocking syscall

This is not true. I can demonstrate:

1) Build this PR on Linux, and run `mono/tests/bug-80307.exe` with that Mono
2) Comment out the `install`+`uninstall`
3) Run the test again; the test will hang.

The particular test just happens to exit without shutting down its FSW, causing this situation.

Reply:
@luhenry I see from `suspend_sync` that what you are saying maybe should happen if `interrupt_kernel` is true, however you can observe that this is not happening. 

Even if the runtime did deliver a signal automatically, this is still not enough. We need to be able to distinguish between a runtime abort interrupt, and a kernel EINTR that just wants us to do a loop. I do not believe there is a way to distinguish this in managed code.

Reply:
You should look more closely at how and where this interrupt mechanism is used: it's used by Thread.Abort and Thread.Interrupt to be able to force the thread to exit a blocking syscall which wouldn't otherwise be interrupted by a signal. It's not here to replace the interrupting property of a signal (or the OS API we rely on), it's here to **complement** it. So sending a signal (or the OS API we rely on) **in** the interrupt callback is just double work.

Now, how does Thread.Interrupt or Thread.Abort then use this interrupt token/callback: 1. it will send the signal (or the OS API we rely on) to suspend the thread, 2. set a flag on the thread so that we throw a `ThreadAbortException` or a `ThreadInterruptedException` based on what you tried to do, and 3. use the interrupt token to call the interrupt callback (ie `interrupt_read` in this case). This "flag on the thread" is then checked in and out of internal call transitions, so that if the thread was interrupted, it can throw the proper exception.

That explanation aside, what you have currently doesn't make the difference between a EINTR which happened because of an explicit interrupt or a spurious interruption of the syscall from the point of view of managed since you are not returning "interrupted" in any case.

The gist of it is: you can safely do the `while` loop in managed without installing/uninstalling the token since it's all redundant to what the runtime is already doing.

Reply:
Thank you for the explanation, I see all of it is handled! I copied the interrupt callback from `ves_icall_System_Net_Sockets_Socket_Accept_internal`, which does use `mono_threads_suspend_abort_syscall` in the callback, but that situation must be different somehow.

> That explanation aside, what you have currently doesn't make the difference between a EINTR which happened because of an explicit interrupt or a spurious interruption of the syscall from the point of view of managed since you are not returning "interrupted" in any case.


I am using `interrupted`: 
```
} while (!interrupted && result < 0 && errno == EINTR);
```

> The gist of it is: you can safely do the while loop in managed without installing/uninstalling the token since it's all redundant to what the runtime is already doing.

I will do that :)

Reply:
Let me take a look at Accept to understand better why it's needed.

For the `interrupted`, you only use it to know if you should break the loop because you don't check if the "flag on the thread" is set. If you switch the loop to managed, the "flag on the thread" will be checked when transitioning from managed to runtime.